Indi controller

EwoudSmeur opened the pull request:
*Not intended to merge*

I just made this pull request in order to review the code and have my comments in one place.

whoenig committed:
Support to read latest logging file from uSD card over radio

This change adds a new memory type (USD), which maps to the last
logging file on the microSD card. The file can be read over the
radio if logging is disabled (i.e., usd.logging is set to 0).

Tested with downloadUSDLogfile from crazyflie_tools with and
without the uSD-card deck using a CF2.0.

whoenig committed:
uSD logging: Fix issue with empty log files.

If logging is enabled/disabled at runtime, sometimes empty log
files were created. This adds an additional check to only write
a new file if logging is enabled.

whoenig committed:
Implement full pose measurement in EKF.

This adds two new messages to the crtp_localization_service to
receive external pose information (compressed and not compressed)
such as information obtained by a motion capture system. (Fixes
issue #203).

It also adds a full pose update in the EKF using a scalar update
for both position and orientation errors. (Fixes issue #246).

Tested with the Crazyswarm using a single Crazyflie and a VICON
motion capture system.

krichardsson committed:
Merge pull request #445 from whoenig/feature_ekf_pose_update

Implement full pose measurement in EKF.

krichardsson committed:
Merge branch 'master' into featureUSDdownload

krichardsson committed:
Merge pull request #438 from whoenig/featureUSDdownload

Support to read latest logging file from uSD card over radio

Commit:
Adding new controller possibility

Commit:
Adding new controller possibility

Commit:
Adding new controller possibility - INDI

Commit:
Adding new controller possibility - INDI

Commit:
Second order Butterworth low pass filter

Commit:
Second order Butterworth low pass filter

knmcguire committed:
#447 added /doc folder and config files

knmcguire committed:
#447 added an simple layout with index and several pages

knmcguire committed:
#446 changed config file jykell

knmcguire committed:
#447 update a file id

knmcguire committed:
#447 update config.yml

knmcguire committed:
#447 moved .md pages

knmcguire committed:
#447 update config file

knmcguire committed:
#447 changed menu name to handle with the relative name bug

knmcguire committed:
Merge branch 'master' of https://github.com/bitcraze/crazyflie-firmware

Commit:
INDI possibility

Commit:
INDI possibility

Commit:
Removing the integer implementation and the fourth order filter

Commit:
Removing the integer implementation and the fourth order filter

krichardsson committed:
Added log variable for motor/propeller test results. Closes #448

krichardsson committed:
#448 Corrected motor test variances

Commit:
correcting license

Commit:
correcting license

Commit:
First steps of the INDI controller

Commit:
First steps of the INDI controller

Commit:
Avoid macros for things that can be an inline function

Commit:
Avoid macros for things that can be an inline function

Commit:
Calculate the desired angular acceleration

Commit:
Calculate the desired angular acceleration

Commit:
INDI feedback

Commit:
INDI feedback

Commit:
INDI feedback

Commit:
INDI feedback

EwoudSmeur commented on src/modules/interface/controller_indi.h:
```diff
@@ -0,0 +1,125 @@
+
+#define STABILIZATION_INDI_ESTIMATION_FILT_CUTOFF 4.0f
+
+#define PERIODIC_FREQUENCY 512.0f
```
The frequency of the controller is specified somewhere else in the Crazyflie code, please use that value

andre-paraense replied:
As far as I understood, we should instead be using the `ATTITUDE_RATE` specified in the `stabilizer_types.h`, which corresponds to the IMU frequency. Is that correct?

EwoudSmeur replied:
I think that is correct. Can you take a look what the function of this line is?
https://github.com/amreelab/crazyflie-firmware/blob/indi-controller/src/modules/src/controller_mellinger.c#L133

andre-paraense replied:
It seems to be testing the `ATTITUDE_RATE` (500Hz) in relation to the `MAIN_LOOP_RATE` (1000Hz), trying to guarantee that the `tick` is a multiple of this relation in order to keep running the controller. Do you understand what is the reason of this test? Should we be doing this test too?

EwoudSmeur replied:
https://github.com/amreelab/crazyflie-firmware/blob/indi-controller/src/modules/src/stabilizer.c#L210

It think that answers your question 😉

andre-paraense replied:
Inserted this verification to skip calls faster than ATTITUDE_RATE.

EwoudSmeur commented on src/modules/interface/controller_indi.h:
```diff
@@ -0,0 +1,125 @@
+#define STABILIZATION_INDI_REF_ERR_R 600.0f
+#define STABILIZATION_INDI_REF_RATE_P 28.0f
+#define STABILIZATION_INDI_REF_RATE_Q 28.0f
+#define STABILIZATION_INDI_REF_RATE_R 28.0f
```
These parameters are surely too aggressive for the Crazyflie, try to start with 100.0f and 14.0f for ERR and RATE respectively

andre-paraense replied:
Done! I still do not have this sensibility, I lack symbol grounding in the real world for these values. I think as soon as we start experiment I will have a better notion.

EwoudSmeur replied:
STABILIZATION_INDI_REF_RATE_P this one is the amount of angular acceleration in rad/s2 that you want per rad/s rate error. The other one is actually based on a quaternion error. Here is the quaternion for a certain rotation theta:
![image](https://user-images.githubusercontent.com/2464596/62368199-3bebac00-b52c-11e9-9263-10cdfe23ef2b.png)
Which will give a value in u_x etc that is approx. half as big. So the other gain, STABILIZATION_INDI_REF_ERR_R, should be halved if we feed back the angles instead. But then it is [gain] (rad/s ) / rad of attitude error.

Actually, for this latter one, the implementation matters:
1) ang_acc_desired = P*att_err - D* rate
2) ang_acc_desired = D*(P*att_err - rate)
That gain of 100 that I mentioned was actually based on method 1, with quaternion feedback. If we use method 2 with angle feedback, we have to do 100/D/2 = 100/14/2 = 3.57

For first order actuator dynamics A(s) = 1/(tau*s+ 1) with tau = 1/16, from experience and looking at the poles of the total transfer function I know that these gains should perform well. Also consider section 2.1 in "Cascaded Incremental Nonlinear Dynamic Inversion Control for MAV Disturbance Rejection"

Sorry for the lengthy explanation and giving you wrong numbers at first.

andre-paraense replied:
So in the end we should be using 3.57f for ERR and 14.0f for RATE respectively?

EwoudSmeur replied:
Yes, if angles are in radians

EwoudSmeur commented on src/modules/interface/controller_indi.h:
```diff
@@ -0,0 +1,125 @@
+#define STABILIZATION_INDI_FULL_AUTHORITY false
+#define STABILIZATION_INDI_ACT_DYN_P 0.1f
+#define STABILIZATION_INDI_ACT_DYN_Q 0.1f
+#define STABILIZATION_INDI_ACT_DYN_R 0.1f
```
This we should determine with flight data, or better with an actuator identification setup. It is the discrete coefficient, so it actually depends on the control frequency. I think these actuators will have a (continuous) first order time constant of around 1/16, you can find the discrete value with a zero order hold with the correct control frequency.

andre-paraense replied:
Did you mean we will calculate these constants with flight data, with the help of the SD-Card, or that we can calculate them with the control frequency? I got confused here, sorry.

EwoudSmeur replied:
I mean we should determine these dynamics, but until we do, we can assume (based on experience) A(s) = 1/(tau*s+ 1) with tau=1/16. To get the discrete version of this with a zero order hold, you have to know ATTITUDE_RATE. For instance with Matlab you can do c2d(A, 1/ATTITUDE_RATE).

andre-paraense replied:
```
A = control.TransferFunction([1],[1/16,1])
Ad = control.matlab.c2d(A,1/500)

> Ad

0.03149
----------
z - 0.9685

dt = 0.002
```

Is this what you meant?

EwoudSmeur replied:
Yes!

andre-paraense replied:
Nice.

However, I still do not understand pragmatically what we can do with this discrete version in hands, or if/how it may help us determine the dynamics parameter.

EwoudSmeur replied:
`#define STABILIZATION_INDI_ACT_DYN_R 0.03149f` 😉

andre-paraense replied:
Alright, got it! :+1:

EwoudSmeur commented on src/modules/src/controller_indi.c:
```diff
@@ -0,0 +1,311 @@
+  float tau_r = 1.0f / (2.0f * PI * STABILIZATION_INDI_FILT_CUTOFF_R);
+  float tau_axis[3] = {tau, tau, tau_r};
+  float tau_est = 1.0f / (2.0f * PI * STABILIZATION_INDI_ESTIMATION_FILT_CUTOFF);
+  float sample_time = 1.0f / PERIODIC_FREQUENCY;
```
You can see that for this discrete filter, the control frequency is taken into account. Therefore, the cutoff can be specified independent of the control frequency. For the actuator dynamics I should also do that.

andre-paraense replied:
Did not understand if action is needed here or not.

EwoudSmeur replied:
Nope 😉

EwoudSmeur commented on src/modules/src/controller_indi.c:
```diff
@@ -0,0 +1,311 @@
+	 //Don't increment if thrust is off
+	 //TODO: this should be something more elegant, but without this the inputs
+	 //will increment to the maximum before even getting in the air.
+	 if(control->thrust == 0) {
```
Will the setpoint thrust be exactly zero, or can there be a slight bias?
The point here is that the the INDI controller can integrate very fast, and if the vehicle is standing on the ground, giving low thrust, giving a roll command will not make the vehicle rotate. The roll command can therefore quickly accumulate, and if you then add thrust you take off with a huge roll command. Turning off the control until you are ready to take off is a bit of a dirty remedy.
On the other hand, this should not be triggered during flight, as it will turn off the control. Therefore the threshold should be kept low.
Also, please use the setpoint for this, and not the control varbiable.

andre-paraense replied:
Got it! Changed inspired by the Paparazzi code.

EwoudSmeur commented on src/modules/src/controller_indi.c:
```diff
@@ -0,0 +1,311 @@
+	float_rates_zero(&indi.u_in);
+
+	// Re-initialize filters
+	indi_init_filters();
```
Can this also be called during flight, for instance when switching controllers? Then the filters should not be reset to zero but to the current values of sensors and actuators.

andre-paraense replied:
I have no idea if the can be called during flight. I am not sure if controller can be changed real time like that. Will leave a `TODO` in the code to check this later.

EwoudSmeur replied:
ok!

andre-paraense replied:
Here is an evidence that the controller can be changed dynamically:

- https://github.com/amreelab/crazyflie-firmware/blob/996b1fdcb60847d4cf0bbba238685b23e35204bf/src/modules/src/stabilizer.c#L267

I will change the code as you suggested ;)

andre-paraense replied:
How are we supposed to set the filters to current values of sensors and actuators in this INIT method if we do not have access to their values (sensors and actuators values)? Maybe instead of resetting to zero we should just not set the values at all??

EwoudSmeur replied:
Good point... We would need to propose to send the state and last control signals along in the init function. Maybe write it down somewhere, it is not essential for your first flight if you only use INDI. But if we want to support switching between controllers in flight it should be implemented.

The best thing would even be if the filters were always running, such that there are no transient effects, but I think for such changes we need to acquire some goodwill ;)

EwoudSmeur commented on src/modules/src/controller_indi.c:
```diff
@@ -0,0 +1,311 @@
+	 */
+
+	float stateAttitudeRateRoll = radians(sensors->gyro.x);
+	float stateAttitudeRatePitch = -radians(sensors->gyro.y);
```
Are you sure about this minus?

andre-paraense replied:
Yes, I am sure. There is this caveat in the Crazyflie that the Y Axis is inverted. It is documented and used like that in all of their controllers. It is something we have to be aware of all the time not to forget.

EwoudSmeur replied:
ok!

EwoudSmeur commented on src/modules/src/controller_indi.c:
```diff
@@ -0,0 +1,311 @@
+	 * though this will be inaccurate for large attitude errors, but it will be ok for now.
+	 * 4.2 Angular_acceleration_reference = D * (rate_reference – rate_measurement)
+	 */
+	 float referenceAttitudeRateRoll = radians(setpoint->attitudeRate.roll);
```
Is the desired roll rate already calculated somewhere else?

andre-paraense replied:
I think so yes. Both PID and Mellinger controllers use this value like that. Check it and see if you agree:
- https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/controller_mellinger.c#L254
- https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/controller_pid.c#L64

EwoudSmeur replied:
Ok, there should be a gain on the attitude error that produces the desired rate, it would be good if you can find that such that we can change its value if needed.

andre-paraense replied:
Maybe here?

https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/attitude_pid_controller.c#L115

andre-paraense replied:
If we follow on this path, I believe we end up with the gains declared in `pid.h`:
https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/interface/pid.h#L33

EwoudSmeur replied:
But do we need to call a PID function that uses these gains, or is that already done outside of our INDI controller file? If you are comparing with the gains that I gave, take into account that I was using radians and quaternions, while here sometimes I see radians and sometimes degrees.
If you'll use this, consider that we don't need integrator gains as INDI is already sort of an integrator. I also think it doesn't make sense to have overlapping gains (e.g. D gain for the attitude as well as a P gain for the rates).

andre-paraense replied:
As far as I understand, it is already done outside of our INDI controller. As the Mellinger controller only use radians as well, and we are using these values as they come in the same way the Mellinger is using, I think we have a tested proof that our approach works regarding units (radians x degrees). Take a look at how Mellinger does here:
https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/controller_mellinger.c#L250
Maybe what you are trying to tell me is that we will not be able to use these setpoints as they come from the PID rate controller, but we will have to recalculate them. Is that so?

EwoudSmeur commented on src/modules/src/controller_indi.c:
```diff
@@ -0,0 +1,311 @@
+
+	  //bound the total control input
+	if(STABILIZATION_INDI_FULL_AUTHORITY){
+	  clamp(indi.u_in.p, -9600.0f, 9600.0f);
```
Here, use the maximum and minimum allowable command as specified for Crazyflie.

andre-paraense replied:
Could not find this specification, but used some values inspired in this Mellinger controller code:
- https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/controller_mellinger.c#L289

EwoudSmeur reviewed (commented):
Hi @andre-paraense , looks good, but consider the specific case of Crazyflie in terms of control frequency, control input limits, etc.

Commit:
Frequency of the controller as the IMU fequency

Commit:
Frequency of the controller as the IMU fequency

Commit:
Adapting parameters to Crazyflie reality

Commit:
Adapting parameters to Crazyflie reality

Commit:
Increment after in the air

Commit:
Increment after in the air

Commit:
TODO to check if controller can be changed dynamically during flight

Commit:
TODO to check if controller can be changed dynamically during flight

Commit:
maximum and minimum allowable command as specified for Crazyflie

Commit:
maximum and minimum allowable command as specified for Crazyflie

whoenig committed:
Fix issue in locSrv that can cause incorrect decoding of external
position or pose estimates.

whoenig committed:
Fix compressState acceleration values to be mm/s^2.

whoenig committed:
Add support to control the yaw angle (and not only rate).

Note that stabilizationModeYaw is in the firmware already, but
was not used until now.

krichardsson committed:
Merge pull request #451 from whoenig/bugfix-yawAngleCtrl

Add support to control the yaw angle (and not only rate).

krichardsson committed:
Merge pull request #450 from whoenig/bugfix-compressState

Fix compressState acceleration values to be mm/s^2.

krichardsson committed:
Merge pull request #449 from whoenig/bugfix-locSrv

Fix issue in locSrv that can cause incorrect decoding of external

Commit:
Skipping calls faster than ATTITUDE_RATE

Commit:
Skipping calls faster than ATTITUDE_RATE

Commit:
Correcting parameters for ERR and RATE

Commit:
Correcting parameters for ERR and RATE

Commit:
Correcting parameters for ERR and DYN

Commit:
Correcting parameters for ERR and DYN

knmcguire committed:
[docs] converted wiki pages to markdown and made it compatible with githubpages

knmcguire committed:
[docs] further simplifying for githubpages

Commit:
Logging and adjusting thrust initial level

Commit:
Logging and adjusting thrust initial level

knmcguire committed:
[docs] included some kind of logical structure for the documentation

knmcguire committed:
[docs] fixed markdown typo in build instructions

Commit:
logs and parameters

Commit:
logs and parameters

Commit:
shorten parms and log names

Commit:
shorten parms and log names

krichardsson committed:
#448 Added counter for the number of motor tests that have been run. Replaces the test finished bit in motorPass

krichardsson committed:
#447 Updated docs to new structure

krichardsson committed:
#447 Added menu

Commit:
logging and small bugs

Commit:
logging and small bugs

Commit:
Adaptive INDI as a parameter

Commit:
Adaptive INDI as a parameter

Commit:
Adaptive as default

Commit:
Adaptive as default

Commit:
signal correction

Commit:
signal correction

krichardsson committed:
Added support for %c in printf and fixed length bug for negative ints. Closes #432

knmcguire committed:
#447 added ctrp docs from wiki to this repository

krichardsson committed:
Renamed variables. Closes #425

krichardsson committed:
#364 Wait for system to be fully started before running LED sequences

krichardsson committed:
#216 Added .gitattributes file to make sure future commits use correct line endings

krichardsson committed:
#216 Converted line endings of existing files to UNIX style

knmcguire committed:
#447 adding some extra doc pages on development

krichardsson committed:
#447 Updated internal links and images

krichardsson committed:
#447 More updates of images and internal links

knmcguire committed:
#447 fixed links for deployment (still need to fix overview pages)

Commit:
control effectiveness estimated

Commit:
control effectiveness estimated

Commit:
control effectiveness estimated

Commit:
control effectiveness estimated

krichardsson committed:
#290 Refactored and fixed automatic mode scanning

krichardsson committed:
#447 Updated urls and removed wrap tags

Commit:
control effectiveness estimation for INDI controller

Commit:
control effectiveness estimation for INDI controller

knmcguire committed:
#447 updated image descriptions and checked links

Commit:
script for extracting signals from usd deck card data

Commit:
script for extracting signals from usd deck card data

ntamas committed:
fix compilation of unit tests on macOS

krichardsson committed:
Merge pull request #457 from collmot/fix/unit-tests-on-mac

fix compilation of unit tests on macOS

krichardsson committed:
#457 Added same solution for params

knmcguire committed:
Merge branch 'master' of https://github.com/bitcraze/crazyflie-firmware

knmcguire committed:
#447 added alt attributes to images and fixed linking issues

krichardsson committed:
#458 Added kalman supervisor

krichardsson committed:
Merge branch 'master' of github.com:bitcraze/crazyflie-firmware

NicksonYap committed:
#459 - params for radio link connectivity

NicksonYap committed:
LED Ring - RSSI Effect - turns white if disconnected

Commit:
Parameter estimation

Commit:
Parameter estimation

Commit:
inverting yaw signal, following other impl

Commit:
inverting yaw signal, following other impl

Commit:
taking out estimation for a while

Commit:
taking out estimation for a while

Commit:
Parameter estimation

Commit:
Parameter estimation

Commit:
Improving thrust threshold

Commit:
Improving thrust threshold

Commit:
Improving code

Commit:
Improving code

Commit:
Improving parameters and logging

Commit:
Improving parameters and logging

Commit:
cleaning code

Commit:
cleaning code

Commit:
cleaning unused lms estimation code

Commit:
cleaning unused lms estimation code

Commit:
Cleaning adaptive part, exposing parameters

Commit:
Cleaning adaptive part, exposing parameters

Commit:
correcting first order actuator dynamics constant

Commit:
correcting first order actuator dynamics constant

Commit:
correcting g2 constant

Commit:
correcting g2 constant

Commit:
Improved calculation of the desired angular acceleration

Commit:
Improved calculation of the desired angular acceleration

krichardsson committed:
#464 Added initial version of the Active marker deck driver

krichardsson committed:
#464 Updated i2c address for active marker deck

Commit:
Improving attitude error calculation

Commit:
Improving attitude error calculation

Commit:
improving cycle rates

Commit:
improving cycle rates

Commit:
Parametizing actuator dynamics and filter cutoff

Commit:
Parametizing actuator dynamics and filter cutoff

Commit:
logging mellinger signals

Commit:
logging mellinger signals

Commit:
Caculating attitude error simple way

Commit:
Caculating attitude error simple way

Commit:
logging variables

Commit:
logging variables

krichardsson committed:
#464 Updated how IDs are written to the deck and added test that reads a version string from the deck to make sure it is responding.

NicksonYap committed:
get_bs_position - fix decode("utf-8")

krichardsson committed:
Merge pull request #466 from NicksonYap/led_rssi

LED RSSI - turns white if disconnected (1 sec timeout)

krichardsson committed:
#466 Refactoring

ataffanel committed:
Closes #469: Update lighthouse bitstream to V2

krichardsson committed:
Merge pull request #467 from NicksonYap/lh_fix_get_bs_pos

get_bs_position - fix decode("utf-8")

krichardsson committed:
#464 Added version check of active marker deck firmware and improved start up sequence.

krichardsson committed:
Removed jenkins file

krichardsson committed:
The "buffer full" marker functionality seems to be broken, fixed by rewriting part if it. Buffer full marker added after current message if possible. Incoming data discarded until current message is sent. Closes #471

tobbeanton committed:
Added platform support for Crazyflie Bolt. Closes #472

First renaming from RZR to Bolt. Then the RZR deck driver was removed
and instead the the motor mapping is done via the platform  setup. A
test file for the Bolt was also added.

tobbeanton committed:
Merge branch 'master' of https://github.com/bitcraze/crazyflie-firmware

tobbeanton committed:
Removed bolt.c deck from makefile. #472

tobbeanton committed:
Fixed include problem in test setup. #472

krichardsson committed:
Moved the lighthouse binary to avoid mixup with CF binaries in releases

krichardsson committed:
#472 Reconfigured Bolt, antennas are not physically close

krichardsson committed:
#472 Added motor mapping for Tags

tobbeanton committed:
Changed Z-ranger2 estimator to complementary. Closes #473

The complementry estimator does not track position and is a better
option when flying manually. The vl53l1 sensor update rate was increased
to 40Hz to not get oscilating hight stabilization and distance mode set
to medium to compensate for that.

tobbeanton committed:
Merge branch 'master' of https://github.com/bitcraze/crazyflie-firmware

krichardsson committed:
Corrected copy/paste error in disabled flow deck code. Closes #475

whoenig committed:
Add LED-ring effect to show status of localization service.

krichardsson committed:
Merge pull request #477 from whoenig/feature_locsrv_led

Add LED-ring effect to show status of localization service.

whoenig committed:
Refactor planner to avoid global static members

This simplifies using the same code in other C-projects as well
as creating Python bindings using SWIG. The latter is used in the
Crazyswarm simulator for more accurate (and efficient) simulation.

krichardsson committed:
Merge pull request #478 from whoenig/refactor_planner

Refactor planner to avoid global static members

whoenig committed:
Improved Tumble Detection and Handling

This change improves several issues with the current tumble
detector:
  1. The old detector was only enabled if the Kalman filter
     was initially enabled. If one switched dynamically to a
     different state estimator, this important safety feature did
     not engage.
  2. The old detector triggered by angle, which is an estimated
     value. If the state estimator diverged, no tumble was detected.
     The new solution relies on the accelerometer instead, a direct
     sensor measurement.
  3. The old detector was not sticky, i.e., a user might pick
     up a CF and it starts spinning again, see issue #258. The new
     version triggers an emergencyStop instead. The emergency stop
     can be reset and read using the new parameter stabilizer.stop.
  4. Smaller fixes:
     a) Add more #ifdef guards to only compile actually enabled code.
     b) Interface clean-ups

This replaces the PR #470. This change has been tested on a bench
test and with the Crazyswarm.

whoenig committed:
Avoid compilation error on newer toolchains (2019-q3).

This change temporarily disables a warning (that leads to a
compiler error in release builds) in USB_OTG_WritePacket and
USB_OTG_ReadPacket.

krichardsson committed:
Merge pull request #482 from whoenig/bugfix_compilerError

Avoid compilation error on newer toolchains (2019-q3).

krichardsson committed:
Merge pull request #481 from whoenig/bugfix_issue258

Improved Tumble Detection and Handling

Commit:
Merge branch 'indi-controller' of github.com:amreelab/crazyflie-firmware into indi-controller

andre-paraense commented:
Closing as this was not meant to be merged and all issues have been resolved.
